### PR TITLE
CHANGE(pmd): @W-17310830@: Update pmd-wrapper to run rules by language forcefully

### PR DESCRIPTION
**Pull request by stephen-carter-at-sf**

### CODE REVIEW

File: packages/code-analyzer-pmd-engine/pmd-cpd-wrappers/src/main/java/com/salesforce/sfca/pmdwrapper/PmdRunner.java [hunk 1]
```diff
@@ -0,0 +1,130 @@
...
+
+        // Force the language so that pmd doesn't look at file extensions. Note: we already associated the files based
+        // on their file extensions to the correct languages the typescript side.
+        Language pmdLanguageId = config.getLanguageRegistry().getLanguageById(language);
+        if (pmdLanguageId == null) {
+            throw new RuntimeException("The language \"" + language + "\" is not recognized by PMD.");
+        }
+        LanguageVersion forcedLangVer = config.getLanguageVersionDiscoverer()
+                .getDefaultLanguageVersion(pmdLanguageId);
+        config.setForceLanguageVersion(forcedLangVer);
```
Comment (stephen-carter-at-sf) [inline packages/code-analyzer-pmd-engine/pmd-cpd-wrappers/src/main/java/com/salesforce/sfca/pmdwrapper/PmdRunner.java:56]: This here is the big win. We now can ignore pmd's default association from file extension to language and just force the language ourselves on a given set of files.

So now we can associate on the typescript side the file_extensions that are associated with a specific pmd language to accomodate for the PMD AppExchange Rules, etc.

File: packages/code-analyzer-pmd-engine/pmd-cpd-wrappers/src/main/java/com/salesforce/sfca/pmdwrapper/PmdRunner.java [hunk 2]
```diff
@@ -0,0 +1,130 @@
+package com.salesforce.sfca.pmdwrapper;
```
Comment (stephen-carter-at-sf) [inline packages/code-analyzer-pmd-engine/pmd-cpd-wrappers/src/main/java/com/salesforce/sfca/pmdwrapper/PmdRunner.java:1]: This file was actually just renamed from PmdRuleRunner... not sure why github thought this was a delete and an add. Sorry for the noise.

File: packages/code-analyzer-pmd-engine/pmd-cpd-wrappers/src/main/java/com/salesforce/sfca/shared/ProgressReporter.java [hunk 3]
```diff
@@ -0,0 +1,37 @@
+package com.salesforce.sfca.shared;
```
Comment (stephen-carter-at-sf) [inline packages/code-analyzer-pmd-engine/pmd-cpd-wrappers/src/main/java/com/salesforce/sfca/shared/ProgressReporter.java:1]: Nothing new in this shared folder... just moved stuff. Again sorry for the noise.

File: packages/code-analyzer-pmd-engine/pmd-cpd-wrappers/src/test/java/com/salesforce/sfca/pmdwrapper/PmdWrapperTest.java [hunk 4]
```diff
@@ -126,16 +127,143 @@ void whenCallingMainWithDescribeWithCustomRulesetsFile_thenRulesetsAreApplied(@T
...
+        String inputFile = createTempFile(tempDir, "inputFile.json", inputFileContents);
+
+        String[] args = {"run", inputFile, "/does/not/matter"};
+        RuntimeException thrown = assertThrows(RuntimeException.class, () -> callPmdWrapper(args));
+        assertThat(thrown.getMessage(), is(
+                "Error while attempting to invoke PmdRunner.run: The \"runDataPerLanguage\" field was not set."));
+    }
+
+    @Test
+    void whenCallingRunWithZeroLanguages_thenError(@TempDir Path tempDir) throws Exception {
```
Comment (jfeingold35) [inline packages/code-analyzer-pmd-engine/pmd-cpd-wrappers/src/test/java/com/salesforce/sfca/pmdwrapper/PmdWrapperTest.java]: Nitpick: Could the test name explain what it means to have zero languages, the way `whenCallingRunWithEmptyArrayFor_filesToScan_thenError` does? Maybe `whencallingRunWithEmptyObjectFor_runDataPerLanguage_thenError`? Or is that too wordy to be helpful?

Comment (stephen-carter-at-sf) [reply packages/code-analyzer-pmd-engine/pmd-cpd-wrappers/src/test/java/com/salesforce/sfca/pmdwrapper/PmdWrapperTest.java]: Sure. Fixed.

**Review by jfeingold35 — APPROVED** [review]
